osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Osu.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Osu.Objects;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+{
+    /// <summary>
+    /// Represents the skill required to memorise and hit every object in a map with the Flashlight mod enabled.
+    /// </summary>
+    public class Flashlight : OsuStrainSkill
+    {
+        public Flashlight(Mod[] mods)
+            : base(mods)
+        {
+        }
+
+        protected override double SkillMultiplier => 0.13;
+        protected override double StrainDecayBase => 0.15;
+        protected override double DecayWeight => 1.0;
+        protected override int HistoryLength => 10; // Look back for 10 notes is added for the sake of flashlight calculations.
+
+        protected override double StrainValueOf(DifficultyHitObject current)
+        {
+            if (current.BaseObject is Spinner)
+                return 0;
+
+            var osuCurrent = (OsuDifficultyHitObject)current;
+            var osuHitObject = (OsuHitObject)(osuCurrent.BaseObject);
+
+            double scalingFactor = 52.0 / osuHitObject.Radius;
+            double smallDistNerf = 1.0;
+
+            double result = 0.0;
+
+            if (Previous.Count > 0)
+            {
+                double cumulativeStrainTime = 0.0;
+
+                for (int i = 0; i < Previous.Count; i++)
+                {
+                    var osuPrevious = (OsuDifficultyHitObject)Previous[i];
+                    var osuPreviousHitObject = (OsuHitObject)(osuPrevious.BaseObject);
+
+                    if (!(osuPrevious.BaseObject is Spinner))
+                    {
+                        double jumpDistance = (osuHitObject.StackedPosition - osuPreviousHitObject.EndPosition).Length;
+
+                        cumulativeStrainTime += osuPrevious.StrainTime;
+
+                        // We want to nerf objects that can be easily seen within the Flashlight circle radius.
```

Comment:
Shouldn't we also look ahead and nerf circles that are visible but aren't gonna be hit next? Ideally even check if any circles are visible along a perfect cursor path.
Another idea if you implement things above is to calculate how long every circle been visible before its hit and apply nerf based on that. If you tie HD into this you might not need arbitrary HDFL bonus at all

Reply:
This is sort of the point of the system, to nerf patterns which have some circles that are easily visible (such as back and forth jumps). However I know there are other cases (such as circles which are visible along a perfect cursor path, like you said, but are far away from previous circles), but implementing this would require a lot more changes especially since I would probably need to account for follow points, approach circles and hit circles and it seems out of the scope of the proposal for now.

As for HD, I can't think of another way to do that than to make another difficulty skill for HD+FL. Although there might be a better way and I'm just not familiar enough with the code to know